### PR TITLE
Bridge: Add channel redirection

### LINKED ISSUE #14560: Redirecting channel output to another channel
There are situations where it could be useful for data sent by a channel to be directly received by another channel instead of being sent to the client. This applies particularly to cases where a larger amount of data is processed, e.g.:

- Downloading a file and saving it on the server or uploading a file from the server.
- Copying files from one server to another. (This would be useful for someone implementing a file manager plugin.)
- Live migration of Podman containers between servers. (That's what I'm working on in cockpit-podman.)

I tried implementing this for local channels using a `redirect` option with the target channel id and it seems to work well with http-stream2, fsread1 and fsreplace1. I also tested redirecting from a peer and successfully copied a file from a remote host to the one on which Cockpit runs. If everything goes well, I should have a PR ready soon.

### PR DESCRIPTION
**Pull request by lenticularis39**
This adds a "redirect" option to the open command, which takes the name
of the channel the output of the new one will be redirected to. The
redirect is effectively applied as close to the channels as possible.

Closes #14560

---

There are two cases in which the output is redirected from a channel:
- The output is redirected from a local channel. This is handled in `CockpitRouter` by setting the `redirect_target` property of `CockpitChannel`, which uses it to send data to the redirect target instead of the transport in `cockpit_channel_actual_send`.
- The output is redirected from a peer. This is handled in `cockpit_peer_handle` by passing the redirect target as an argument and then by putting the redirect into a hash table with the (source) channel name as a key. `on_other_recv` then looks up the channel name in the table, again sending it to the redirect target instead of the transport if a corresponding entry is present.

Similarly there are two redirect target types: a local channel and a peer. These are handled by different implementations of the `CockpitRedirect` interface - `CockpitChannelRedirect` for local channels (holding a reference to the corresponding `CockpitChannel` object) and `CockpitPeerRedirect` for peers (holding a reference to the transport and the channel name).

Redirection of a channel is attempted on all bridges - the shortest path then takes precedence (e.g. if there is a redirect between two channels on the same peer, it happens on both the parent bridge and the peer, but effectively the data is redirected on the peer and nothing arrives on the parent). This also means it is compatible with older versions of cockpit-bridge - they simply ignore the redirect option and the redirect is done on the parent bridge (this can obviously be slower, but it will work).

I hope this makes sense - it seems like a small, but useful feature to me. (For potential uses see the linked issue.)

### CODE REVIEW

**Comment by martinpitt** [comment]
This is a really cool patch, thanks @lenticularis39 ! Honestly I'm quite impressed how clean this already is (with tests and all).

I (or someone else) will soon review this in detail, but in the meantime, can you please fix the build? (See Semaphore failure):

```
src/common/cockpitredirect.c: In function ‘cockpit_channel_redirect_interface_init’:
src/common/cockpitredirect.c:65:15: error: assignment to ‘void (*)(CockpitRedirect *, GBytes *)’ {aka ‘void (*)(struct _CockpitRedirect *, struct _GBytes *)’} from incompatible pointer type ‘void (*)(CockpitChannelRedirect *, GBytes *)’ {aka ‘void (*)(struct _CockpitChannelRedirect *, struct _GBytes *)’} [-Werror=incompatible-pointer-types]
   iface->send = cockpit_channel_redirect_send;
               ^
src/common/cockpitredirect.c: In function ‘cockpit_peer_redirect_interface_init’:
src/common/cockpitredirect.c:147:15: error: assignment to ‘void (*)(CockpitRedirect *, GBytes *)’ {aka ‘void (*)(struct _CockpitRedirect *, struct _GBytes *)’} from incompatible pointer type ‘void (*)(CockpitPeerRedirect *, GBytes *)’ {aka ‘void (*)(struct _CockpitPeerRedirect *, struct _GBytes *)’} [-Werror=incompatible-pointer-types]
   iface->send = cockpit_peer_redirect_send;
               ^
```

File: src/bridge/cockpitrouter.c
Comment (martinpitt) [inline src/bridge/cockpitrouter.c:764]: Can you please add this to doc/authentication.md? This documents the JSON wire protocol, standard options (like this one), and the payload specific ones.

I'm not sure whether this should really be called "redirect", it feels more like an Unix pipe. But that's bikeshedding. :-)

Comment (lenticularis39) [reply src/bridge/cockpitrouter.c:764]: > I'm not sure whether this should really be called "redirect", it feels more like an Unix pipe. But that's bikeshedding. :-)

I didn't want to call it pipe because of possible confusion with `CockpitPipe`. (I guess it could be named "channel pipe", but  `CockpitChannelRedirect` already has its specific meaning.)

Comment (lenticularis39) [reply src/bridge/cockpitrouter.c:764]: > Can you please add this to doc/authentication.md? This documents the JSON wire protocol, standard options (like this one), and the payload specific ones.

I assume you meant doc/protocol.md (which contains the items you mentioned).

Comment (martinpitt) [reply src/bridge/cockpitrouter.c:764]: Err yes, of course. :blush:

**Comment by martinpitt** [comment]
@lenticularis39 : Thanks! That fixed the build, but there is still a memory leak (semaphore failure):
```
PASS: test-router 15 /router/superuser/start
==729== 6 bytes in 1 blocks are definitely lost in loss record 15 of 1,459
==729==    at 0x10083874F: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==729==    by 0x100ACB4E0: g_malloc (gmem.c:99)
==729==    by 0x100AE51BE: g_strdup (gstrfuncs.c:363)
==729==    by 0x40A66D: test_external_channel_redirect (test-router.c:330)
==729==    by 0x100AED39D: test_case_run (gtestutils.c:2597)
==729==    by 0x100AED39D: g_test_run_suite_internal (gtestutils.c:2685)
==729==    by 0x100AED143: g_test_run_suite_internal (gtestutils.c:2697)
==729==    by 0x100AED861: g_test_run_suite (gtestutils.c:2772)
==729==    by 0x100AED880: g_test_run (gtestutils.c:2007)
==729==    by 0x409C36: main (test-router.c:1042)
==729== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: definite
   fun:malloc
   fun:g_malloc
   fun:g_strdup
   fun:test_external_channel_redirect
   fun:test_case_run
   fun:g_test_run_suite_internal
   fun:g_test_run_suite_internal
   fun:g_test_run_suite
   fun:g_test_run
   fun:main
}
[... some more... ]
ERROR: test-router process failed: 33
```

You should be able to reproduce this with

    make check-memory TESTS=test-router

**Comment by lenticularis39** [comment]
I see. I'll have a look at it :)

File: src/bridge/test-router.c
Comment (martinpitt) [inline src/bridge/test-router.c:330]: That's one of the leaks -- [json_object_set_array_member](https://developer.gnome.org/json-glib/stable/json-glib-JSON-Object.html#json-object-set-array-member) does not take ownership of `key`, it dups it internally. So just drop the g_strdup().

**Comment by martinpitt** [comment]
Also, feel free to peruse the new `g_autoptr`, `g_autofree` etc., to simplify cleanup paths and keeping track of allocations. A lot of the existing code does not use this yet (as it's fairly old), but we started using these in newer/updated code.

**Comment by lenticularis39** [comment]
Thank you for the hints! Memory leaks should be fixed now.

**Comment by stefwalter** [comment]
This is fascinating and really cool. But it does cause me to wonder about the use cases?

Does this mean that the cockpit-bridge must stay running throughout the entirety of a long term operation? We typically allow Cockpit to disconnect from a server side service, or process, and expect that process to continue without the user being logged in.

**Comment by martinpitt** [comment]
Right, for long-running use cases we now have a [library and example](https://github.com/cockpit-project/cockpit/tree/master/examples/long-running-process). I suppose this is for operations which don't take *that* long, but where it would still be too expensive to transfer everything through the browser wire back and forth?

I'd like to see a pkg/playground/ page which utilizes this for something simple, like combining a http with a file write channel ("direct file download").This would provide a hands-on example and an integration test. I'm even tempted to write this myself, I just currently lack some time for this. But this doesn't need to block this PR.

**Comment by lenticularis39** [comment]
> I suppose this is for operations which don't take _that_ long, but where it would still be too expensive to transfer everything through the browser wire back and forth?

The original idea was transferring data between two Cockpit sessions on different hosts, where the connection between them would likely be much faster than the connection to the client (and the same applies for local redirection usage, e.g. downloading a file). Of course the connection between the hosts (or the Cockpit session on one host) depends on the client being connected, so it's not efficient for long term operations.

**Review by stefwalter — CHANGES_REQUESTED** [review]
I love this work! But before going into further review of specifics, I believe there's two high level things that need to be addressed.

**1 CockpitRedirectChannel does not respect channel freezes.**

Freezing is implemented in CockpitTransport on received messages, for a given channel. I believe it is necessary to send all redirected messages through CockpitTransport logic, even those that are local. This also makes the structure of the classes more simple and easy to understand:

 * Turn ```CockpitRedirect``` into a single class that always acts on a ```transport``` + ```channel```.
 * Add an boolean called ```self->inbound``` to ```CockpitRedirect```
 * If ```self->inbound``` is ```TRUE``` then messages are routed into the ```CockpitTransport``` via ```cockpit_transport_emit_recv``` ... and if ```self->inbound``` is ```FALSE``` then use ```cockpit_transport_send```.

What do you think about this change? Am I missing an aspect you already considered? 

**2 There does not seem to be any code dealing with channel closure.**

As you know from Unix sockets, there is an EPIPE state ... where the destination of a redirect closes and no longer accepts data, therefore the sender has to error out. Is it possible to fix this, given the architecture of Cockpit?

File: src/bridge/cockpitpeer.c
Comment (lenticularis39) [reply src/bridge/cockpitpeer.c:1011]: I agree, this behavior could be quite confusing.

**Comment by lenticularis39** [comment]
> **1 CockpitRedirectChannel does not respect channel freezes.**
> 
> Freezing is implemented in CockpitTransport on received messages, for a given channel. I believe it is necessary to send all redirected messages through CockpitTransport logic, even those that are local. This also makes the structure of the classes more simple and easy to understand:
> 
>     * Turn `CockpitRedirect` into a single class that always acts on a `transport` + `channel`.
> 
>     * Add an boolean called `self->inbound` to `CockpitRedirect`
> 
>     * If `self->inbound` is `TRUE` then messages are routed into the `CockpitTransport` via `cockpit_transport_emit_recv` ... and if `self->inbound` is `FALSE` then use `cockpit_transport_send`.
> 
> 
> What do you think about this change? Am I missing an aspect you already considered?

Thank you for the suggestion. After looking into the Cockpit source code I think it makes much more sense to redirect the messages through the transport rather than through the channel, not only because of channel freezes, but also for other features that may be implemented in `CockpitTransport` in the future. (Also there might be other callback functions connected to the transport `recv` signal than the `CockpitChannel`'s `on_transport_recv`.)

I'll change the `CockpitRedirect` logic to reflect this as you suggested.

> **2 There does not seem to be any code dealing with channel closure.**
> 
> As you know from Unix sockets, there is an EPIPE state ... where the destination of a redirect closes and no longer accepts data, therefore the sender has to error out. Is it possible to fix this, given the architecture of Cockpit?

That's certainly a good idea. I believe this can be implemented - `CockpitRedirect` can track whether the channel is closed or not by connecting a handler to the `control` and `closed` signals of the `CockpitTransport` (in case of an outbound redirect) and to the `closed` signal of the `CockpitChannel` class (in case of an inbound redirect). The latter would require to keep the distinction between `CockpitChannelRedirect` and `CockpitPeerRedirect`, though, because the `CockpitChannel` instance would have to be passed to the class for the purpose of connecting (and disconnecting) of the signal.

**Comment by stefwalter** [comment]
So it appears there is one very fundamental limitation. 

The cockpit protocol operates across a tree. A tree that has directionality:

 * Inbound (ie: toward the leaves, "bridges", the various CockpitRouter instances), 
 * Outbound (ie: toward the trunk, cockpit-ws and the browser)

```
   cockpit-bridge        cockpit-bridge
    --privileged          --privileged
       \                       /
        \                     /
         \             cockpit-bridge
          \                /
      cockpit-bridge -----/
              \
               \
               |
            cockpit-ws
               |
               |
            cockpit.js
            
```

I don't think this implementation can facilitate redirection between things that are on different "branches" here. For example the two privileged bridges (in this example these are running on two different hosts.

At first I thought this was a corner case, but in practice such complex protocol trees are used routinely.

@lenticularis39 Am I understanding things right?

@martinpitt WDYT?

**Comment by stefwalter** [comment]
Perhaps implementing this at the CockpitRouter level would help avoid this limitation? It appears that would mean inspecting outbound messages though.

**Comment by lenticularis39** [comment]
I believe this works. Let's say I want to redirect from channel "a" on the left privileged bridge (let's call it bridge 1) to channel "b" on the second privileged bridge (let's call it bridge 2). 

I open channel "b" through an open command - the command is forwarded to bridge 2 (through `process_open_peer` in `CockpitRouter`), which opens the channel and tracks the path to it in the `channel_peers` tables inside `CockpitRouter`s in all the bridges in the middle (the tracking was added by the PR).

Then I open channel "a", passing the "redirect" option with value "b" to the open command. Redirect is then attempted on all bridges the command passes through by looking up channel "b" in both the `CockpitRouter::channels` and `CockpitRouter::channel_peers`. Here only the parent bridge knows about channel "b", so the redirect succeeds there, passing all data arriving from the next bridge (on the path to bridge 1; in this case it's bridge 1 itself, but it doesn't have to be) on channel "a" to channel "b" (this is handled in `CockpitPeer::on_other_recv`).

Also see `test-router.c` - at the end of `test_external_channel_redirect` there is a test that redirects the output of one external channel to another external channel.

**Comment by stefwalter** [comment]
You're right. That's what I get for doing code review before drinking my Coffee. :frowning_face:

**Comment by lenticularis39** [comment]
Changes:

- `cockpit_peer_handle` and `CockpitChannel` now ref the `CockpitRedirect` object instead of taking ownership of it.
-  `CockpitChannelRedirect` now forwards data to the underlying transport via `cockpit_transport_emit_recv` instead of using `cockpit_channel_internal_recv`.
- Both implementations of `CockpitRedirect` track the state of the redirect target by registering signal handlers to the channel/transport, marking it as closed when the channel or the underlying transport closes. If sending data through a redirect is attempted and the target is closed, `cockpit_redirect_send` returns FALSE, triggering the closing of the source channel with problem `redirect-target-closed`.
- Closing of redirect target was added to the tests.

Limitations:

- The client has to make sure the redirect target is open when the second channel is opened. `CockpitRedirect` doesn't know that, because the handlers are registered only after creating it.

**Comment by stefwalter** [comment]
Hi @lenticularis39 ... sorry for the delay.

There was one aspect of this that was missing. We should treat redirect channels that do not (or no longer) exist as closed channels. This is because Cockpit heavily relies on ordering guarantees in order to assure correctness.

Thus if a channel has already closed, then the redirecting channel should get a "redirect-closed". The same applies if the channel never existed. The two states are indistinguishable. 

I pushed a FIXUP commint that implements this change. WDYT?

**Comment by lenticularis39** [comment]
Hi @stefwalter,

Unfortunately this breaks the redirect mechanism, because the bridge where the redirecting channel is created doesn't have to know about the redirect channel, e.g. when redirecting a superuser channel to a non-superuser one (the first one is on a child bridge that doesn't know about the channels on its parent). In these cases a bridge closes the channel, although the redirect succeeded on its predecessor in the graph.

I believe this could be solved by dropping the `redirect-target` argument in the forwarded open command in such cases (they can be detected by looking up the redirect channel in the `channels` attribute of `CockpitPeer`). I'll try to implement this and also write a test case that checks it - apparently I missed this case when writing the tests.

**Comment by lenticularis39** [comment]
So it turned out that the only reason the issue didn't appear in tests was the fact that they use `mock-bridge` as the peer, which doesn't use `CockpitRouter`.

Anyway I implemented the fix and it seems to work properly now.

**Comment by stefwalter** [comment]
I'm trying to understand the security implications of this. There is a bit of a caveat where connections to different hosts are not strictly part of the same login session, and allowing arbitrary redirection from one to the other doesn't quite fit into the security model of Cockpit.

Related to the problem we never fixed in https://github.com/cockpit-project/cockpit/pull/6260

@martinpitt @mvollmer I'd like to talk with you about this before merging. The security model on multi-host is lacking (which is why we hadn't shipped multi-host support enabled by default on RHEL).

**Comment by martinpitt** [comment]
Sorry for dropping the ball on this one for so long. I don't want to postpone this forever, but indeed we need to understand the implications. So far I had the impression that this is just a "fast path" optimization, but does not allow anything new that the JavaScript side could not already do. I.e. right now, injected malicious JavaScript can already open channels to all logged in remote sessions, and do `.read()`, `.modify()` on them -- the data just needs to be funnelled to and from the browser.

@stefwalter , @lenticularis39 , did I misunderstand this? Could the "automatic" routing somehow "accidentally" pick a remote channel where it shouldn't?

**Comment by stefwalter** [comment]
That's correct, although it encourages behavior that we may later prohibit.

Given: Web Browser -> Host A -> Host B

Currently we don't encourage Host B to perform actions to Host A, although (unfortunately) the security model allows it. This code encourages that behavior. If we intend to shore up security so that Host B cannot access Host A indiscriminately, then we shouldn't encourage that behaviour now.

**Comment by lenticularis39** [comment]
> So far I had the impression that this is just a "fast path" optimization, but does not allow anything new that the JavaScript side could not already do. I.e. right now, injected malicious JavaScript can already open channels to all logged in remote sessions, and do `.read()`, `.modify()` on them -- the data just needs to be funnelled to and from the browser.

That's correct - code from any host (remote or local) can access the other hosts by opening channels on any host simply by setting the `host` parameter in the `open` command or interfere with existing channels by simply sending a message on them (knowing their channel id) - remote channel redirect only adds another way how this can be done.

> Could the "automatic" routing somehow "accidentally" pick a remote channel where it shouldn't?

That can only happen if the JavaScript code sends a remote channel ID as the redirect target by mistake. If I understand it correctly channel IDs are unique, therefore a channel ID used for a local channel can't be used for a remote channel during a single session.

**Comment by lenticularis39** [comment]
> That's correct, although it encourages behavior that we may later prohibit.

I would say that it not only encourages it, but it doesn't make much sense without it. For example if I want to copy a file from one host to another during a multi-machine action (like in cockpit-project/cockpit-podman#590) I have to open the channel that will save the file on the other host, too.

> Currently we don't encourage Host B to perform actions to Host A, although (unfortunately) the security model allows it. This code encourages that behavior. If we intend to shore up security so that Host B cannot access Host A indiscriminately, then we shouldn't encourage that behaviour now.

The first question is whether multi-machine operations (as mentioned in #14560; like the aforementioned Cockpit container migration, a file manager supporting copying files between machines and so on) are something that should be included in Cockpit and/or Cockpit modules. If not, the remote channel redirect feature probably wouldn't have any use and should be removed (and my Cockpit container migration PR closed, despite being almost complete, since it wouldn't be something fitting in Cockpit). If yes (which I initially assumed from how remote hosts are implemented in Cockpit), then the security problems would have to be resolved in a way allowing enabling host-to-host communication when desired - here I don't see any difference between accessing another host directly and through a channel redirection from the perspective of the security fix implementation.

**Comment by stefwalter** [comment]
> The first question is whether multi-machine operations (as mentioned in #14560; like the aforementioned Cockpit container migration, a file manager supporting copying files between machines and so on) are something that should be included in Cockpit and/or Cockpit modules.

We need to figure out how to model the security. Without Cockpit, when one connects to Host A with SSH and then from there again uses ssh to connect to Host B:

SSH -> Host A -> Host B

The Linux user does not expect that Host B can access anything that the ssh process on Host A has access to. Currently in Cockpit Host B *can* access Host A.

 * Either we have to model this security behavior in the Cockpit UX, so that the user understands the implications, despite it being different from typical expectations of Linux login sessions. 

or:

 * Disallow and prevent this access, thus matching the user's typical expectations for login sessions.

Personally, I'm 50/50 on this, and I'd like to understand where other maintainers in the project think we should go.

**Comment by allisonkarlitskaya** [comment]
I think I'm really the wrong person for reviewing this PR, although I can think of one scenario where it might be useful which doesn't violate any assumptions about the security model: bastion host.

A user might login from a web browser via bastion host to any number of servers in a cluster and may meaningfully want to migrate things between them, being OK with it going via the bastion host (since we're probably all on a local network here) and gaining a huge advantage from it not having to bubble all the way up to the web browser (possibly at home, on wireless, or even on a mobile network).

**Comment by martinpitt** [comment]
I'm revisiting old PRs. I'm sorry that this didn't work out, especially for all the time that you sank into it! Now we are about to move to the Python re-implementation of the bridge, so we won't make big changes to the C bridge any more. If you are still interested in this feature, can we do a video call to design this up-front?

**Comment by lenticularis39** [comment]
I did this work as part of an internship project (on Cockpit container migration, see also cockpit-project/cockpit-podman#590), which was concluded after this PR got stuck on the above security issue.

I didn't close the PR, because there was some chance I could continue working on it for my master's thesis or another kind of school project, but I decided go on a different path. Hence I won't be working on this feature anymore.

**Pull request closed without merging**
